File: packages/js/core/src/models/resnet.ts
import { Tensor } from '../ir/tensor.js';
import { recordOp } from '../macros.js';
const add = (a: Tensor, b: Tensor) => recordOp('Add', [a, b]);
const globalAveragePool = (x: Tensor) => recordOp('GlobalAveragePool', [x]);
const maxPool = (x: Tensor, attr?: any) => recordOp('MaxPool', [x], attr);
const flatten = (x: Tensor) => recordOp('Flatten', [x]);

import { BatchNormalization, ConvND, Gemm, Relu } from '../primitives.js';

function getParam(name: string, shape: number[], dtype: any = 'float32'): Tensor {
  return new Tensor(name, shape, dtype, false, false, new Float32Array());
}

export class BasicBlock {
  static expansion: number = 1;
  public prefix: string;
  public conv1: ConvND;
  public bn1: BatchNormalization;
  public relu: Relu;
  public conv2: ConvND;
  public bn2: BatchNormalization;
  public downsample: boolean;
  public downsampleConv?: ConvND;
  public downsampleBn?: BatchNormalization;

  constructor(
    inplanes: number,
    planes: number,
    stride: number = 1,
    downsample: boolean = false,
    prefix: string = '',
  ) {
    this.prefix = prefix;
    this.conv1 = new ConvND(2, inplanes, planes, 3, stride, 1, 1, 1, false);
    this.bn1 = new BatchNormalization(planes);
    this.relu = new Relu();
    this.conv2 = new ConvND(2, planes, planes, 3, 1, 1, 1, 1, false);
    this.bn2 = new BatchNormalization(planes);
    this.downsample = downsample;

    if (this.downsample) {
      this.downsampleConv = new ConvND(
        2,
        inplanes,
        planes * BasicBlock.expansion,
        1,
        stride,
        0,
        1,
        1,
        false,
      );
      this.downsampleBn = new BatchNormalization(planes * BasicBlock.expansion);
    }
  }

  call(x: Tensor): Tensor {
    let identity = x;

    let out = this.conv1.call(
      x,
      getParam(`${this.prefix}.conv1.weight`, [
        this.conv1.outChannels,
        this.conv1.inChannels,
        3,
        3,
      ]),
    );
    out = this.bn1.call(
      out,
      getParam(`${this.prefix}.bn1.weight`, [this.bn1.numFeatures]),
      getParam(`${this.prefix}.bn1.bias`, [this.bn1.numFeatures]),
      getParam(`${this.prefix}.bn1.running_mean`, [this.bn1.numFeatures]),
      getParam(`${this.prefix}.bn1.running_var`, [this.bn1.numFeatures]),
    );
    out = this.relu.call(out);

    out = this.conv2.call(
      out,
      getParam(`${this.prefix}.conv2.weight`, [
        this.conv2.outChannels,
        this.conv2.inChannels,
        3,
        3,
      ]),
    );
    out = this.bn2.call(
      out,
      getParam(`${this.prefix}.bn2.weight`, [this.bn2.numFeatures]),
      getParam(`${this.prefix}.bn2.bias`, [this.bn2.numFeatures]),
      getParam(`${this.prefix}.bn2.running_mean`, [this.bn2.numFeatures]),
      getParam(`${this.prefix}.bn2.running_var`, [this.bn2.numFeatures]),
    );

    if (this.downsample && this.downsampleConv && this.downsampleBn) {
      identity = this.downsampleConv.call(
        x,
        getParam(`${this.prefix}.downsample.0.weight`, [
          this.downsampleConv.outChannels,
          this.downsampleConv.inChannels,
          1,
          1,
        ]),
      );
      identity = this.downsampleBn.call(
        identity,
        getParam(`${this.prefix}.downsample.1.weight`, [this.downsampleBn.numFeatures]),
        getParam(`${this.prefix}.downsample.1.bias`, [this.downsampleBn.numFeatures]),
        getParam(`${this.prefix}.downsample.1.running_mean`, [this.downsampleBn.numFeatures]),
        getParam(`${this.prefix}.downsample.1.running_var`, [this.downsampleBn.numFeatures]),
      );
    }

    // We use a dummy add operation that returns a generic Tensor for now
    // Assuming add is in registry or primitives
    // Let's implement dummy fallback
    const recordOp = (opType: string, inputs: Tensor[], attr?: any) => {
      const dtype = inputs[0]?.dtype ?? 'float32';
      return new Tensor(`${opType}_out`, [], dtype, false, false, new Float32Array());
    };

    out = recordOp('Add', [out, identity]);
    out = this.relu.call(out);

    return out;
  }
}

export class ResNet {
  public inplanes: number = 64;
  public conv1: ConvND;
  public bn1: BatchNormalization;
  public relu: Relu;
  public layer1: BasicBlock[];
  public layer2: BasicBlock[];
  public layer3: BasicBlock[];
  public layer4: BasicBlock[];
  public fc: Gemm;
  public numClasses: number;

  constructor(layers: number[], numClasses: number = 1000) {
    this.conv1 = new ConvND(2, 3, this.inplanes, 7, 2, 3, 1, 1, false);
    this.bn1 = new BatchNormalization(this.inplanes);
    this.relu = new Relu();

    this.layer1 = this.makeLayer(64, layers[0]!, 1, 'layer1');
    this.layer2 = this.makeLayer(128, layers[1]!, 2, 'layer2');
    this.layer3 = this.makeLayer(256, layers[2]!, 2, 'layer3');
    this.layer4 = this.makeLayer(512, layers[3]!, 2, 'layer4');

    this.fc = new Gemm(1.0, 1.0, 0, 1); // transB = 1
    this.numClasses = numClasses;
  }

  private makeLayer(
    planes: number,
    blocks: number,
    stride: number = 1,
    prefix: string = '',
  ): BasicBlock[] {
    let downsample = false;
    if (stride !== 1 || this.inplanes !== planes * BasicBlock.expansion) {
      downsample = true;
    }

    const layers: BasicBlock[] = [];
    layers.push(new BasicBlock(this.inplanes, planes, stride, downsample, `${prefix}.0`));
    this.inplanes = planes * BasicBlock.expansion;
    for (let i = 1; i < blocks; i++) {
      layers.push(new BasicBlock(this.inplanes, planes, 1, false, `${prefix}.${i}`));
    }

    return layers;
  }

  call(x: Tensor): Tensor {
    x = this.conv1.call(
      x,
      getParam('conv1.weight', [this.conv1.outChannels, this.conv1.inChannels, 7, 7]),
    );
    x = this.bn1.call(
      x,
      getParam('bn1.weight', [this.bn1.numFeatures]),
      getParam('bn1.bias', [this.bn1.numFeatures]),
      getParam('bn1.running_mean', [this.bn1.numFeatures]),
      getParam('bn1.running_var', [this.bn1.numFeatures]),
    );
    x = this.relu.call(x);

    const recordOp = (opType: string, inputs: Tensor[], attr?: any) => {
      const dtype = inputs[0]?.dtype ?? 'float32';
      return new Tensor(`${opType}_out`, [], dtype, false, false, new Float32Array());
    };

    x = recordOp('MaxPool', [x], { kernel_shape: [3, 3], strides: [2, 2], pads: [1, 1, 1, 1] });

    for (const layer of [this.layer1, this.layer2, this.layer3, this.layer4]) {
      for (const block of layer) {
        x = block.call(x);
      }
    }

    x = recordOp('GlobalAveragePool', [x]);
    x = recordOp('Flatten', [x]);

    const fcW = getParam('fc.weight', [this.numClasses, 512 * BasicBlock.expansion]);
    const fcB = getParam('fc.bias', [this.numClasses]);

    x = this.fc.call(x, fcW, fcB);

    return x;
  }
}

export function resnet18(numClasses: number = 1000): ResNet {
  return new ResNet([2, 2, 2, 2], numClasses);
}

export function resnet50(numClasses: number = 1000): ResNet {
  return new ResNet([3, 4, 6, 3], numClasses);
}

File: packages/python/onnx9000-core/src/onnx9000/core/models/resnet.py
"""Module docstring."""

import logging
from collections.abc import Sequence
from typing import Any, Callable, Optional, Union

from onnx9000.core.ir import Tensor, Variable
from onnx9000.core.ops import add, flatten, global_average_pool, max_pool
from onnx9000.core.primitives import BatchNormalization, ConvND, Gemm, Relu


def get_param(name: str, shape: list[int], dtype: int = 1) -> Variable:  # noqa: D103
    return Variable(name=name, shape=shape, dtype=dtype)


class BasicBlock:  # noqa: D101
    expansion: int = 1

    def __init__(  # noqa: D107
        self,
        inplanes: int,
        planes: int,
        stride: int = 1,
        downsample: bool = False,
        prefix: str = "",
    ):
        self.prefix = prefix
        self.conv1 = ConvND(
            2, inplanes, planes, kernel_size=3, stride=stride, padding=1, bias=False
        )
        self.bn1 = BatchNormalization(planes)
        self.relu = Relu()
        self.conv2 = ConvND(2, planes, planes, kernel_size=3, stride=1, padding=1, bias=False)
        self.bn2 = BatchNormalization(planes)
        self.downsample = downsample
        if self.downsample:
            self.downsample_conv = ConvND(
                2, inplanes, planes * self.expansion, kernel_size=1, stride=stride, bias=False
            )
            self.downsample_bn = BatchNormalization(planes * self.expansion)

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        identity = x

        out = self.conv1(
            x,
            get_param(
                f"{self.prefix}.conv1.weight",
                [self.conv1.out_channels, self.conv1.in_channels, 3, 3],
            ),
        )
        out = self.bn1(
            out,
            get_param(f"{self.prefix}.bn1.weight", [self.bn1.num_features]),
            get_param(f"{self.prefix}.bn1.bias", [self.bn1.num_features]),
            get_param(f"{self.prefix}.bn1.running_mean", [self.bn1.num_features]),
            get_param(f"{self.prefix}.bn1.running_var", [self.bn1.num_features]),
        )
        out = self.relu(out)

        out = self.conv2(
            out,
            get_param(
                f"{self.prefix}.conv2.weight",
                [self.conv2.out_channels, self.conv2.in_channels, 3, 3],
            ),
        )
        out = self.bn2(
            out,
            get_param(f"{self.prefix}.bn2.weight", [self.bn2.num_features]),
            get_param(f"{self.prefix}.bn2.bias", [self.bn2.num_features]),
            get_param(f"{self.prefix}.bn2.running_mean", [self.bn2.num_features]),
            get_param(f"{self.prefix}.bn2.running_var", [self.bn2.num_features]),
        )

        if self.downsample:
            identity = self.downsample_conv(
                x,
                get_param(
                    f"{self.prefix}.downsample.0.weight",
                    [self.downsample_conv.out_channels, self.downsample_conv.in_channels, 1, 1],
                ),
            )
            identity = self.downsample_bn(
                identity,
                get_param(f"{self.prefix}.downsample.1.weight", [self.downsample_bn.num_features]),
                get_param(f"{self.prefix}.downsample.1.bias", [self.downsample_bn.num_features]),
                get_param(
                    f"{self.prefix}.downsample.1.running_mean", [self.downsample_bn.num_features]
                ),
                get_param(
                    f"{self.prefix}.downsample.1.running_var", [self.downsample_bn.num_features]
                ),
            )

        out = add(out, identity)
        out = self.relu(out)

        return out


class ResNet:
    """ResNet implementation built using IR macros/primitives."""

    def __init__(self, block_type: type, layers: list[int], num_classes: int = 1000):  # noqa: D107
        self.inplanes = 64
        self.conv1 = ConvND(2, 3, self.inplanes, kernel_size=7, stride=2, padding=3, bias=False)
        self.bn1 = BatchNormalization(self.inplanes)
        self.relu = Relu()

        self.layer1 = self._make_layer(block_type, 64, layers[0], prefix="layer1")
        self.layer2 = self._make_layer(block_type, 128, layers[1], stride=2, prefix="layer2")
        self.layer3 = self._make_layer(block_type, 256, layers[2], stride=2, prefix="layer3")
        self.layer4 = self._make_layer(block_type, 512, layers[3], stride=2, prefix="layer4")

        self.fc = Gemm(trans_b=1)  # PyTorch linear uses transB
        self.num_classes = num_classes

    def _make_layer(
        self, block_type: type, planes: int, blocks: int, stride: int = 1, prefix: str = ""
    ) -> list[Any]:
        downsample = False
        if stride != 1 or self.inplanes != planes * block_type.expansion:
            downsample = True

        layers = []
        layers.append(block_type(self.inplanes, planes, stride, downsample, prefix=f"{prefix}.0"))
        self.inplanes = planes * block_type.expansion
        for i in range(1, blocks):
            layers.append(block_type(self.inplanes, planes, prefix=f"{prefix}.{i}"))

        return layers

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        # Initial Convolution
        x = self.conv1(
            x, get_param("conv1.weight", [self.conv1.out_channels, self.conv1.in_channels, 7, 7])
        )
        x = self.bn1(
            x,
            get_param("bn1.weight", [self.bn1.num_features]),
            get_param("bn1.bias", [self.bn1.num_features]),
            get_param("bn1.running_mean", [self.bn1.num_features]),
            get_param("bn1.running_var", [self.bn1.num_features]),
        )
        x = self.relu(x)
        x = max_pool(x, kernel_shape=[3, 3], strides=[2, 2], pads=[1, 1, 1, 1])

        # Layers
        for layer in [self.layer1, self.layer2, self.layer3, self.layer4]:
            for block in layer:
                x = block(x)

        # Classification Head
        x = global_average_pool(x)
        x = flatten(x)

        fc_w = get_param("fc.weight", [self.num_classes, 512 * BasicBlock.expansion])
        fc_b = get_param("fc.bias", [self.num_classes])

        x = self.fc(x, fc_w, fc_b)

        return x


def resnet18(**kwargs: Any) -> ResNet:  # noqa: D103
    return ResNet(BasicBlock, [2, 2, 2, 2], **kwargs)


def resnet50(**kwargs: Any) -> ResNet:  # noqa: D103
    return ResNet(BasicBlock, [3, 4, 6, 3], **kwargs)
